rclpy/src/rclpy/client.cpp

```diff
@@ -0,0 +1,220 @@
+  }
+  auto pycli = py::reinterpret_steal<py::capsule>(pycli_c);
+  // pycli now owns the rclpy_client_t
+  cli.release();
+
+  auto cli_handle = static_cast<rclpy_handle_t *>(pycli);
+  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
+  _rclpy_handle_add_dependency(cli_handle, node_handle);
+  if (PyErr_Occurred()) {
+    _rclpy_handle_dec_ref(cli_handle);
+    throw py::error_already_set();
+  }
+
+  return pycli;
+}
+
+int64_t
+client_send_request(py::capsule pyclient, py::object pyrequest)
+{
+  auto client = static_cast<rclpy_client_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclient.ptr(), "rclpy_client_t"));
+  if (!client) {
+    throw py::error_already_set();
+  }
+
+  destroy_ros_message_signature * destroy_ros_message = NULL;
+  void * raw_ros_request = rclpy_convert_from_py(pyrequest.ptr(), &destroy_ros_message);
+  if (!raw_ros_request) {
+    throw py::error_already_set();
+  }
+
+  int64_t sequence_number;
+  rcl_ret_t ret = rcl_send_request(&(client->client), raw_ros_request, &sequence_number);
+  destroy_ros_message(raw_ros_request);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to send request");
+  }
+
+  return sequence_number;
+}
+
+bool
+client_service_server_is_available(py::capsule pyclient)
+{
+  auto client = static_cast<rclpy_client_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclient.ptr(), "rclpy_client_t"));
+  if (!client) {
+    throw py::error_already_set();
+  }
+
+  bool is_ready;
+  rcl_ret_t ret = rcl_service_server_is_available(client->node, &(client->client), &is_ready);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to check service availability");
+  }
+
+  return is_ready;
+}
+
+py::tuple
+client_take_response(py::capsule pyclient, py::object pyresponse_type)
+{
+  auto client = static_cast<rclpy_client_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclient.ptr(), "rclpy_client_t"));
+  if (!client) {
+    throw py::error_already_set();
+  }
+
+  destroy_ros_message_signature * destroy_ros_message = NULL;
+  void * taken_response_ptr = rclpy_create_from_py(pyresponse_type.ptr(), &destroy_ros_message);
+  if (!taken_response_ptr) {
+    throw py::error_already_set();
+  }
+  auto message_deleter = [destroy_ros_message](void * ptr) {destroy_ros_message(ptr);};
+  auto taken_response = std::unique_ptr<void, decltype(message_deleter)>(
+    taken_response_ptr, message_deleter);
+
+  auto deleter = [](rmw_service_info_t * ptr) {PyMem_Free(ptr);};
+  auto header = std::unique_ptr<rmw_service_info_t, decltype(deleter)>(
+    static_cast<rmw_service_info_t *>(PyMem_Malloc(sizeof(rmw_service_info_t))),
+    deleter);
+  if (!header) {
+    throw std::bad_alloc();
+  }
+
+  rcl_ret_t ret = rcl_take_response_with_info(
+    &(client->client), header.get(), taken_response.get());
+  if (ret == RCL_RET_CLIENT_TAKE_FAILED) {
+    return py::tuple(2);
+  }
+
+  py::tuple result_tuple(2);
+  result_tuple[0] = py::capsule(header.release(), "rmw_service_info_t");
```

Comment:
I think this pycapsule needs a destructor for the `rmw_service_info_t` that calls `PyMem_Free()` since it was `PyMem_Malloc()`'d above.